src/mongocrypt-crypto.c

```diff
@@ -657,47 +681,66 @@ _hmac_step (_mongocrypt_crypto_t *crypto,
       goto done;
    }
 
-   /* [MCGREW]:
-    * """
-    * 4.  The octet string AL is equal to the number of bits in A expressed as a
-    * 64-bit unsigned integer in network byte order.
-    * 5.  A message authentication tag T is computed by applying HMAC [RFC2104]
-    * to the following data, in order:
-    *      the associated data A,
-    *      the ciphertext S computed in the previous step, and
-    *      the octet string AL defined above.
-    * """
-    */
+   /* Construct the input to the HMAC */
+   uint32_t num_intermediates = 0;
+   _mongocrypt_buffer_t intermediates[3];
+   if (AAD &&
+       !_mongocrypt_buffer_from_subrange (
+          &intermediates[num_intermediates++], AAD, 0, AAD->len)) {
+      CLIENT_ERR ("Failed creating MAC view on AD");
+      goto done;
+   }
+   if (!_mongocrypt_buffer_from_subrange (&intermediates[num_intermediates++],
+                                          iv_and_ciphertext,
+                                          0,
+                                          iv_and_ciphertext->len)) {
+      CLIENT_ERR ("Failed creating MAC view on IV and S");
+      goto done;
+   }
 
-   /* Add associated data. */
-   _mongocrypt_buffer_init (&intermediates[0]);
-   _mongocrypt_buffer_init (&intermediates[1]);
-   _mongocrypt_buffer_init (&intermediates[2]);
-   intermediates[0].data = associated_data->data;
-   intermediates[0].len = associated_data->len;
-   /* Add ciphertext. */
-   intermediates[1].data = ciphertext->data;
-   intermediates[1].len = ciphertext->len;
-   /* Add associated data length in bits. */
-   /* multiplying a uint32_t by 8 won't bring it anywhere close to UINT64_MAX */
-   associated_data_len_be = 8 * (uint64_t) associated_data->len;
-   associated_data_len_be = BSON_UINT64_TO_BE (associated_data_len_be);
-   intermediates[2].data = (uint8_t *) &associated_data_len_be;
-   intermediates[2].len = sizeof (uint64_t);
-   tag.data = tag_storage;
-   tag.len = sizeof (tag_storage);
+   // Up in this lexical scope because it needs to live until buffer_concat.
```

Comment:
It's specifically referring to the reference to {AL} which is kept in {intermediates}.
The lifecycle reason for declaring it here rather than in the `if` block seemed inobvious (and indeed, was a briefly a bug source for me).
I've reworded the comment to explain this more clearly.